io.grpc.xds.SharedCallCounterMap.cleanQueue() NullPointerException

Version: v1.38.1
Env: Linux, Jdk8

It happened to grpc client with xds enabled:
```
io.grpc.StatusRuntimeException: INTERNAL: Panic! This is a bug!
        at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:262)
        at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:243)
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:156)
        ......
Caused by: java.lang.NullPointerException: null
        at io.grpc.xds.SharedCallCounterMap.cleanQueue(SharedCallCounterMap.java:76)
        at io.grpc.xds.SharedCallCounterMap.getOrCreate(SharedCallCounterMap.java:67)
        at io.grpc.xds.ClusterImplLoadBalancer.handleResolvedAddresses(ClusterImplLoadBalancer.java:124)
        at io.grpc.util.ForwardingLoadBalancer.handleResolvedAddresses(ForwardingLoadBalancer.java:46)
        at io.grpc.xds.PriorityLoadBalancer$ChildLbState.updateResolvedAddresses(PriorityLoadBalancer.java:269)
        at io.grpc.xds.PriorityLoadBalancer.tryNextPriority(PriorityLoadBalancer.java:136)
        at io.grpc.xds.PriorityLoadBalancer.handleResolvedAddresses(PriorityLoadBalancer.java:102)
        at io.grpc.xds.ClusterResolverLoadBalancer$ClusterResolverLbState.handleEndpointResourceUpdate(ClusterResolverLoadBalancer.java:246)
        at io.grpc.xds.ClusterResolverLoadBalancer$ClusterResolverLbState.access$1600(ClusterResolverLoadBalancer.java:157)
        at io.grpc.xds.ClusterResolverLoadBalancer$ClusterResolverLbState$EdsClusterState$1EndpointsUpdated.run(ClusterResolverLoadBalancer.java:427)
        at io.grpc.SynchronizationContext.drain(SynchronizationContext.java:95)
        at io.grpc.SynchronizationContext.execute(SynchronizationContext.java:127)
        at io.grpc.xds.ClusterResolverLoadBalancer$ClusterResolverLbState$EdsClusterState.onChanged(ClusterResolverLoadBalancer.java:431)
        at io.grpc.xds.ClientXdsClient$ResourceSubscriber.notifyWatcher(ClientXdsClient.java:1562)
        at io.grpc.xds.ClientXdsClient$ResourceSubscriber.onData(ClientXdsClient.java:1515)
        at io.grpc.xds.ClientXdsClient.handleResourcesAccepted(ClientXdsClient.java:1379)
        at io.grpc.xds.ClientXdsClient.handleEdsResponse(ClientXdsClient.java:982)
        at io.grpc.xds.AbstractXdsClient$AbstractAdsStream.handleRpcResponse(AbstractXdsClient.java:504)
        at io.grpc.xds.AbstractXdsClient$AdsStreamV3$1$1.run(AbstractXdsClient.java:663)
        at io.grpc.SynchronizationContext.drain(SynchronizationContext.java:95)
        at io.grpc.SynchronizationContext.execute(SynchronizationContext.java:127)
        at io.grpc.xds.AbstractXdsClient$AdsStreamV3$1.onNext(AbstractXdsClient.java:655)
        at io.grpc.xds.AbstractXdsClient$AdsStreamV3$1.onNext(AbstractXdsClient.java:652)
        at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onMessage(ClientCalls.java:465)
        at io.grpc.internal.DelayedClientCall$DelayedListener.onMessage(DelayedClientCall.java:447)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInternal(ClientCallImpl.java:652)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:637)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        ... 1 common frames omitted
```

I failed to find the exact steps to reproduce this exception. But it almost always happens in my long-running server application.

--- a/xds/src/main/java/io/grpc/xds/SharedCallCounterMap.java
+++ b/xds/src/main/java/io/grpc/xds/SharedCallCounterMap.java
@@ -52,19 +52,28 @@ final class SharedCallCounterMap implements CallCounterProvider {
 
   @Override
   public synchronized AtomicLong getOrCreate(String cluster, @Nullable String edsServiceName) {
-    cleanQueue();
+    AtomicLong counter = null;
     Map<String, CounterReference> clusterCounters = counters.get(cluster);
+    if (clusterCounters != null) {
+      CounterReference ref = clusterCounters.get(edsServiceName);
+      if (ref != null) {
+        counter = ref.get();
+      }
+    }
+    // In the case of ref.get() == null, must call cleanQueue() prior to creating a new map entry,
+    // otherwise the new entry will be deleted be any cleanQueue().
+    cleanQueue();
+    if (counter != null) {
+      return counter;
+    }
+    clusterCounters = counters.get(cluster);
     if (clusterCounters == null) {
       clusterCounters = new HashMap<>();
       counters.put(cluster, clusterCounters);
     }
-    CounterReference ref = clusterCounters.get(edsServiceName);
-    AtomicLong counter;
-    if (ref == null || (counter = ref.get()) == null) {
-      counter = new AtomicLong();
-      ref = new CounterReference(counter, refQueue, cluster, edsServiceName);
-      clusterCounters.put(edsServiceName, ref);
-    }
+    counter = new AtomicLong();
+    CounterReference ref = new CounterReference(counter, refQueue, cluster, edsServiceName);
+    clusterCounters.put(edsServiceName, ref);
     return counter;
   }
 

--- a/xds/src/main/java/io/grpc/xds/SharedCallCounterMap.java
+++ b/xds/src/main/java/io/grpc/xds/SharedCallCounterMap.java
@@ -58,8 +58,14 @@ final class SharedCallCounterMap implements CallCounterProvider {
       counters.put(cluster, clusterCounters);
     }
     CounterReference ref = clusterCounters.get(edsServiceName);
-    AtomicLong counter;
-    if (ref == null || (counter = ref.get()) == null) {
+    AtomicLong counter = null;
+    if (ref != null) {
+      counter = ref.get();
+      if (counter == null) {
+        ref.enqueue();
+      }
+    }
+    if (counter == null) {
       counter = new AtomicLong();
       ref = new CounterReference(counter, refQueue, cluster, edsServiceName);
       clusterCounters.put(edsServiceName, ref);

--- a/xds/src/main/java/io/grpc/xds/SharedCallCounterMap.java
+++ b/xds/src/main/java/io/grpc/xds/SharedCallCounterMap.java
@@ -61,7 +61,7 @@ final class SharedCallCounterMap implements CallCounterProvider {
       }
     }
     // In the case of ref.get() == null, must call cleanQueue() prior to creating a new map entry,
-    // otherwise the new entry will be deleted be any cleanQueue().
+    // otherwise the new entry will be deleted by any later cleanQueue().
     cleanQueue();
     if (counter != null) {
       return counter;

--- a/xds/src/main/java/io/grpc/xds/SharedCallCounterMap.java
+++ b/xds/src/main/java/io/grpc/xds/SharedCallCounterMap.java
@@ -52,28 +52,19 @@ final class SharedCallCounterMap implements CallCounterProvider {
 
   @Override
   public synchronized AtomicLong getOrCreate(String cluster, @Nullable String edsServiceName) {
-    AtomicLong counter = null;
     Map<String, CounterReference> clusterCounters = counters.get(cluster);
-    if (clusterCounters != null) {
-      CounterReference ref = clusterCounters.get(edsServiceName);
-      if (ref != null) {
-        counter = ref.get();
-      }
-    }
-    // In the case of ref.get() == null, must call cleanQueue() prior to creating a new map entry,
-    // otherwise the new entry will be deleted by any later cleanQueue().
-    cleanQueue();
-    if (counter != null) {
-      return counter;
-    }
-    clusterCounters = counters.get(cluster);
     if (clusterCounters == null) {
       clusterCounters = new HashMap<>();
       counters.put(cluster, clusterCounters);
     }
-    counter = new AtomicLong();
-    CounterReference ref = new CounterReference(counter, refQueue, cluster, edsServiceName);
-    clusterCounters.put(edsServiceName, ref);
+    CounterReference ref = clusterCounters.get(edsServiceName);
+    AtomicLong counter;
+    if (ref == null || (counter = ref.get()) == null) {
+      counter = new AtomicLong();
+      ref = new CounterReference(counter, refQueue, cluster, edsServiceName);
+      clusterCounters.put(edsServiceName, ref);
+    }
+    cleanQueue();
     return counter;
   }
 
@@ -82,6 +73,9 @@ final class SharedCallCounterMap implements CallCounterProvider {
     CounterReference ref;
     while ((ref = (CounterReference) refQueue.poll()) != null) {
       Map<String, CounterReference> clusterCounter = counters.get(ref.cluster);
+      if (clusterCounter.get(ref.edsServiceName) != ref) {
+        continue;
+      }
       clusterCounter.remove(ref.edsServiceName);
       if (clusterCounter.isEmpty()) {
         counters.remove(ref.cluster);

--- a/xds/src/main/java/io/grpc/xds/SharedCallCounterMap.java
+++ b/xds/src/main/java/io/grpc/xds/SharedCallCounterMap.java
@@ -52,6 +52,7 @@ final class SharedCallCounterMap implements CallCounterProvider {
 
   @Override
   public synchronized AtomicLong getOrCreate(String cluster, @Nullable String edsServiceName) {
+    cleanQueue();
     Map<String, CounterReference> clusterCounters = counters.get(cluster);
     if (clusterCounters == null) {
       clusterCounters = new HashMap<>();
@@ -64,7 +65,6 @@ final class SharedCallCounterMap implements CallCounterProvider {
       ref = new CounterReference(counter, refQueue, cluster, edsServiceName);
       clusterCounters.put(edsServiceName, ref);
     }
-    cleanQueue();
     return counter;
   }
 

--- a/xds/src/test/java/io/grpc/xds/SharedCallCounterMapTest.java
+++ b/xds/src/test/java/io/grpc/xds/SharedCallCounterMapTest.java
@@ -62,4 +62,22 @@ public class SharedCallCounterMapTest {
     map.cleanQueue();
     assertThat(counters).isEmpty();
   }
+
+  @Test
+  public void gcAndRecreate() {
+    @SuppressWarnings("UnusedVariable") // assign to null for GC only
+    AtomicLong counter = map.getOrCreate(CLUSTER, EDS_SERVICE_NAME);
+    final CounterReference ref = counters.get(CLUSTER).get(EDS_SERVICE_NAME);
+    assertThat(counter.get()).isEqualTo(0);
+    counter = null;
+    GcFinalization.awaitDone(new FinalizationPredicate() {
+      @Override
+      public boolean isDone() {
+        return ref.isEnqueued();
+      }
+    });
+    map.getOrCreate(CLUSTER, EDS_SERVICE_NAME);
+    assertThat(counters.get(CLUSTER)).isNotNull();
+    assertThat(counters.get(CLUSTER).get(EDS_SERVICE_NAME)).isNotNull();
+  }
 }

